net: lib: call socket poll if could not send data completely in http or ws (websocket)

AndreyDodonov-EH opened the pull request:
In http_client, if we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket.

It's especially relevant after adding https://github.com/zephyrproject-rtos/zephyr/pull/45626

AndreyDodonov-EH commented:
Should we also instead websockets and mqtt (did I miss some others?) with poll?

rlubos commented:
> Should we also extend websockets and mqtt (did I miss some others?) with poll?

MQTT lib handles socket on its own and by design works with blocking socket. As for websocket lib though, it seems to take a socket from the app, so it should also be ready to work with non-blocking.

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -38,6 +38,18 @@ static int sendall(int sock, const void *buf, size_t len)
 			return -errno;
 		}
 
+		if (out_len == 0 && len != 0) {
```
Not sure I understand this check, `out_len == 0` means peer closed the connection, so there's no need to `poll()`. Also, `len` is subtracted after this check, so it might be the case that we've already sent all we wanted, but still issue the `poll()` call. 

IMO it'd be better just to call `zsock_poll()` before `zsock_send()`, unconditionally. We should then check the poll result, if the `POLLOUT` was actualy set on a socket (we could timeout). Finally, the error check from `zsock_send()` should verify if the `errno` was set to `EAGAIN` - it should not report an error in such case but just try again.

AndreyDodonov-EH replied:
>out_len == 0 means peer closed the connection,   

As far as I understand [the socket send spec](https://man7.org/linux/man-pages/man2/send.2.html) , 0 is the number of bytes we could send  (so none), or timeout indication. This was my assumption. 
Is it different in Zephyr? 

>  it might be the case that we've already sent all we wanted
I call poll only if `out_len `is 0, so subtracting 0 from `len `would not mean that we already send anything

AndreyDodonov-EH replied:
Calling zsock_poll _before_ would mean that we call every time, so slowing down sending if we have a lot of small requests, no?

rlubos replied:
Ok, you're right about the first thing, I mixed up return codes with `recv()`, for `send()` 0 doesn't really mean anything else than nothing was send. It is quite unusual though, as a typical return code for a non-blocking socket when it cannot send a data is `-1` with errno set to `EAGAIN`. So if you really want to call `poll()` after initial `send()` I think you should rather check for that, otherwise you'll end up busy looping anyway.

And just FYI, a timeout on a blocking socket, in case `SO_SNDTIMEO` is configured, is also notified by EAGAIN (https://linux.die.net/man/7/socket).

AndreyDodonov-EH replied:
https://linux.die.net/man/7/socket
> if no data has been transferred and the timeout has been reached then -1 is returned with errno set to EAGAIN or EWOULDBLOCK,

From https://man7.org/linux/man-pages/man2/send.2.html 
> EAGAIN or EWOULDBLOCK
              The socket is marked nonblocking and the requested
              operation would block.  POSIX.1-2001 allows either error
              to be returned for this case, and does not require these
              constants to have the same value, so a portable
              application should check for both possibilities.

So  we should probably check `errno `(as you suggest) and out_len as well:
`(errno == -EAGAIN || out_len == -EAGAIN)`

rlubos replied:
In no circumstance an error code is returned directly by a socket call - in case of error, it always return -1 and `errno` value is set accordingly (check "Return Value" section in the send doc). So the check before `poll()` should look like (note that errno value is positive):
`(out_len < 0 && errno == EAGAIN)`

AndreyDodonov-EH replied:
>POSIX.1-2001 allows either error
to be returned for this case, and does not require these
constants to have the same value

This contradicts the fact that socket always returns -1, but if Zephyr does that all the time and never explicit error code, then it's fine.

As for the check, what I observed during busy looping with non-blocking socket, was that out_len was continuously returning 0, again and again.
So we should include 0 for non-blocking case, no?
```
if (out_len <=0 && errno == EAGAIN) { 
   perform polling
} else if (out_len <0) { // as it was 
return -errno; 
}
```

AndreyDodonov-EH replied:
And good point about `EAGAIN` instead `-EAGAIN`, I oversaw it, thanks

rlubos replied:
> out_len was continuously returning 0, again and again.

That's quite surprising, I'd really expect EAGAIN to be returned. But I guess you're right, we need to take 0 into account as well.

> if (out_len <=0 && errno == EAGAIN) { 
>    perform polling
> } else if (out_len <0) { // as it was 
> return -errno; 
> }

If `out_len == 0` we shouldn't check errno, as it shouldn't be set in such case, so I'd say the condition should look like:
```
if (out_len == 0 || (out_len < 0 && errno == EAGAIN) {
    /* poll */
} else if (out_len < 0) {
    return -errno;
}
```

AndreyDodonov-EH replied:
Yes, looks great.
This seems to cover all cases - first part of OR for non-blocking and second part of OR for blocking.

AndreyDodonov-EH replied:
I'll push correspondent changes.

AndreyDodonov-EH commented on subsys/net/lib/http/http_client.c:
```diff
@@ -38,6 +38,18 @@ static int sendall(int sock, const void *buf, size_t len)
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS);
+			if (pollres > 0) {
```
@rlubos What would you suggest on 0 (timeout) and -1 (errors)?
I'd say in case of timeout we proceed (just fall through) and in case of -1 return `errno`?

rlubos replied:
Sounds fine.

AndreyDodonov-EH commented:
@rlubos 
Should I push same websocket adjustment in the same PR or create a new one?

rlubos commented:
> @rlubos Should I push same websocket adjustment in the same PR or create a new one?

It doesn't make a difference for me as long as the commits are separate.

rlubos commented on subsys/net/lib/http/http_client.c:
```diff
@@ -34,7 +34,19 @@ static int sendall(int sock, const void *buf, size_t len)
 		ssize_t out_len = zsock_send(sock, buf, len, 0);
 
-		if (out_len < 0) {
+		if ((out_len == 0 && len != 0) || (out_len < 0 && errno == EAGAIN)) {
```
`len != 0` check seems pointless here - if it wasn't true the loop wouldn't execute in the first place.

rlubos reviewed (commented):
Looks good, thanks, but please squash the `http` related commits - no point introducing a change just to modify it in the next commit. Ideally we have 2 commits here, one for http, the other for websocket.

AndreyDodonov-EH committed:
net: lib: http: call socket poll for http_client send

If we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket.

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>

AndreyDodonov-EH committed:
net: lib: websocket: call socket poll for websocket

If we couldn't send all (or any data) via the socket,
invoke poll instead of blindly retrying and flooding the socket.

Signed-off-by: Andrey Dodonov <Andrey.Dodonov@endress.com>

jukkar commented on subsys/net/lib/http/http_client.c:
```diff
@@ -34,7 +34,19 @@ static int sendall(int sock, const void *buf, size_t len)
+
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS);
```
The `http_client_req()` API call has timeout option, and I think we should take that value into account here. So if the caller has set a timeout then we could do `req->internal.timeout` to wait max amount of time. If user has not set timeout (so that we do not wait), then we should not do any wait here either. Also the timeout should count both send and recv parts, so if we wait here some time, then the wait time in recv should be smaller.

AndreyDodonov-EH replied:
Currently, timeout is enforced via closing the socket by the timer anyway.
Which will cause `zsock_poll `to return, regardless of its timeout. 

But yes, I agree that it doesn't make any sense to have bigger timeout in zsock_poll, so we could add logic to use smaller of timeouts:

```
int timeout = (req_timeout < CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS) ? req_timeout : CONFIG_HTTP_CLIENT_SOCKET_POLL_TIMEOUT_MS;
pollres = zsock_poll(&pfd, 1, req_timeout);
```

What do you think?

AndreyDodonov-EH replied:
@jukkar Could you please have a look at the point above?
Please let me know if that's what you had it mind

jukkar replied:
Sorry for late reply, I just noticed your question.

What I meant that we should initialize the req_timeout to the value that the user is supplying, and then just decrement it until it is 0. This means that we should never exceed the timeout value that the user has supplied in `http_client_req()` call.

jukkar commented on subsys/net/lib/websocket/websocket.c:
```diff
@@ -531,7 +531,20 @@ static int sendmsg_all(int sock, const struct msghdr *message, int flags)
+
+			pfd.fd = sock;
+			pfd.events = ZSOCK_POLLOUT;
+			pollres = zsock_poll(&pfd, 1, CONFIG_WS_CLIENT_SOCKET_POLL_TIMEOUT_MS);
```
Same comment to websocket version, the `websocket_send_msg()` has also timeout option that we should honor.
